src/device/sound.rs

```diff
@@ -0,0 +1,1563 @@
+        }
+        let request_hdr = VirtIOSndHdr::from(CommandCode::VirtioSndRPcmRelease);
+        let rsp: VirtIOSndHdr = self.request(VirtIOSndPcmHdr {
+            hdr: request_hdr,
+            stream_id,
+        })?;
+        // rsp is just a header, so it can be compared with VirtIOSndHdr
+        if rsp == VirtIOSndHdr::from(RequestStatusCode::VirtioSndSOk) {
+            Ok(())
+        } else {
+            Err(Error::IoError)
+        }
+    }
+
+    /// Start a stream with specified stream ID.
+    pub fn pcm_start(&mut self, stream_id: u32) -> Result {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        let request_hdr = VirtIOSndHdr::from(CommandCode::VirtioSndRPcmStart);
+        let rsp: VirtIOSndHdr = self.request(VirtIOSndPcmHdr {
+            hdr: request_hdr,
+            stream_id,
+        })?;
+        // rsp is just a header, so it can be compared with VirtIOSndHdr
+        if rsp == VirtIOSndHdr::from(RequestStatusCode::VirtioSndSOk) {
+            Ok(())
+        } else {
+            Err(Error::IoError)
+        }
+    }
+
+    /// Stop a stream with specified stream ID.
+    pub fn pcm_stop(&mut self, stream_id: u32) -> Result {
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        let request_hdr = VirtIOSndHdr::from(CommandCode::VirtioSndRPcmStop);
+        let rsp: VirtIOSndHdr = self.request(VirtIOSndPcmHdr {
+            hdr: request_hdr,
+            stream_id,
+        })?;
+        // rsp is just a header, so it can be compared with VirtIOSndHdr
+        if rsp == VirtIOSndHdr::from(RequestStatusCode::VirtioSndSOk) {
+            Ok(())
+        } else {
+            Err(Error::IoError)
+        }
+    }
+
+    /// Transfer PCM frame to device, based on the stream type(OUTPUT/INPUT).
+    /// 
+    /// This is a blocking method that will not return until the audio playback is complete.
+    pub fn pcm_xfer(&mut self, stream_id: u32, frames: &[u8]) -> Result {
+        const U32_SIZE: usize = mem::size_of::<u32>();
+        if !self.set_up {
+            self.set_up();
+            self.set_up = true;
+        }
+        if !self.pcm_parameters[stream_id as usize].setup {
+            warn!("Please set parameters for a stream before using it!");
+            return Err(Error::IoError);
+        }
+        let buffer_size = self.pcm_parameters[stream_id as usize].buffer_bytes as usize;
+        assert!(buffer_size * 2 <= frames.len());
+        let mut buf1 = vec![0; U32_SIZE + buffer_size];
+        let mut buf2 = vec![0; U32_SIZE + buffer_size];
+
+        buf1[..U32_SIZE].copy_from_slice(&stream_id.to_le_bytes());
+        buf1[U32_SIZE..U32_SIZE + buffer_size].copy_from_slice(&frames[..buffer_size]);
+
+        buf2[..U32_SIZE].copy_from_slice(&stream_id.to_le_bytes());
+        buf2[U32_SIZE..U32_SIZE + buffer_size]
+            .copy_from_slice(&frames[buffer_size..buffer_size * 2]);
```

Comment:
You could avoid this extra copy by adding the stream ID and buffer separately to the TX queue. Something like:

```rust
let (first, rest) = frames.split_at(buffer_size);
let token1 = unsafe { self.tx_queue.add(&[&stream_id.to_le_bytes(), first], &mut[&mut output1]) }?;
let token2 = unsafe { self.tx_queue.add(&[&stream_id.to_le_bytes(), &rest[..buffer_size]], &mut[&mut output2]) }?;
```

Reply:
Now I've removed `buf1` and `buf2` to avoid the extra copy.

Reply:
Now I've removed `buf1` and `buf2` to avoid the extra copy.